### ISSUE TITLE
[FEAT] AutoComplete widget

### ISSUE THREAD

**Issue by TomJGooding** [Textualize/textual#3362]
I can't see anything on the roadmap, so are there plans to add a AutoComplete widget to Textual? 

The terms get a bit confusing, but by autocomplete I mean a dynamic dropdown list, like you see pretty much everywhere.

Textual already has a Select widget, Input and now a Command Palette, so it might surprise people that there isn't a widget for this (I am aware of other libraries)

**Comment by davep** [comment]
Like https://github.com/darrenburns/textual-autocomplete?

**Comment by TomJGooding** [comment]
Yes something like lextual-autocomplete. Are there plans to add a widget like this to Textual?

(I am aware of this project and obviously Darren's involvement with Textual, which I hope doesn't make this feature request awkward)

**Comment by willmcgugan** [comment]
We were just discussing that today. I think there should be an autocomplete in the core lib.

**Comment by TomJGooding** [comment]
Thanks Will. That's a bit spooky, I really _do_ spend too much time here if I'm somehow picking up your internal discussions...!

**Comment by willmcgugan** [comment]
@davep please sweep the office for bugs.

**Comment by davep** [comment]
Live scenes from the sweep...

![starship-troopers-shoot](https://github.com/Textualize/textual/assets/28237/0c1d054b-fd6e-4426-83b4-a968eae24db7)

**Comment by TomJGooding** [comment]
The core functionality of an `AutoComplete` widget seems to have already been worked out in the command palette.

I confess I don't fully understand the code yet so perhaps this is naive, but couldn't a good chunk be extracted to this new `AutoComplete` widget (then used to derive the command palette)?

I did spend some time experimenting with this idea, but got a bit stuck understanding the command provider.

**Comment by willmcgugan** [comment]
I think @darrenburns 's textual-autocomplete may become the officially recommended solution.
